src/main/java/util/ApiQuotaManager.java

```diff
@@ -0,0 +1,181 @@
+                                                             Utility.minutesFromNow(info.getNextRefreshInMillisecs()),
+                                                             apiCallsUsedInPreviousRefresh, API_QUOTA_BUFFER,
+                                                             DEFAULT_REFRESH_PERIOD_IN_MINS);
+        this.refreshTimer.restartTimer((int) Utility.minsToSecs(refreshDurationInMinutes));
+        logger.info("Refresh period updated to " + refreshDurationInMinutes
+                + "mins with API calls used in previous refresh cycle is " + apiCallsUsedInPreviousRefresh
+                + ", current API quota is " + info.getRemainingRequests() + " and next API quota top-up in "
+                + info.getNextRefreshInMinutesFromNow() + "mins.");
+    }
+
+    /**
+     * Computes the API calls used in previous refresh.
+     * @param newRemainingApiRequests The number of API requests remaining in the current API quota window.
+     * @return The number of API calls used in previous refresh or apiCallsUsedInPreviousRefresh.
+     *         apiCallsUsedInPreviousRefresh will be returned when newRemainingApiRequests > remainingApiRequests.
+     */
+    private int computeApiCallsUsedInPreviousRefresh(int newRemainingApiRequests) {
+        int apiUsageInRecentRefresh = remainingApiRequests - newRemainingApiRequests;
+
+        if (apiUsageInRecentRefresh < 0) {
+            return apiCallsUsedInPreviousRefresh;
+        }
+
+        return apiUsageInRecentRefresh;
+    }
+
+    /**
+     * Computes the TickerTimer period that is used for refreshing the issues periodically
+     * Assumes future refreshes will take the same number of API calls as the last refresh and finds out the refresh
+     * duration that will spread out the refreshes until the next API quota top up.
+     *
+     * Calculation avoids scheduling a refresh that happens during the same time as the API quota top up.
+     *
+     * PMD warning is suppressed to allow explicit parenthesis.
+     *
+     * @param apiQuota The remaining allowed API requests until the next API quota renewal.
+     *                 Pre-condition: >= 0
+     * @param remainingTimeInMins The remaining time left until the next API quota renewal.
+     *                           Pre-condition: >= 0
+     * @param apiCallsUsedInPreviousRefresh The amount of API used in the previous API refresh.
+     *                                      Pre-condition: >= 0
+     * @param apiQuotaBuffer The amount of API calls that is set aside for manual operations by the user.
+     *                       Pre-condition: >= 0
+     * @param minRefreshPeriod The minimal refresh period that will be used.
+     *                         Recommended value : ApiQuotaManager.DEFAULT_REFRESH_PERIOD_IN_MINS
+     *                         Pre-condition: > 0
+     * @return the computed refresh period or minRefreshPeriod.
+     *          There are 3 conditions that minRefreshPeriod will be used.
```

Comment:
--> `minRefreshPeriod will be returned if one of the following conditions are met:`